src/effast.ml

```diff
@@ -139,8 +145,37 @@ let rec normalize_eff t =
   | Option t' ->
     let t'' = normalize_eff t' in
     Option t''
+  | Ref t' ->
+    let t'' = normalize_eff t' in
+    Ref t''
   | List t' ->
     let t'' = normalize_eff t' in
     List t''
   | Fun (s, _, t) -> Fun (normalize_eff s, no_eff, normalize_eff t)
 ;;
+
+let some typ payload eff = Constructor (typ, "Some", [ payload ], eff)
+let none typ = Constructor (typ, "None", [], (false, false))
+
+module Ref = struct
+  let ref_t =
+    let new_tv = newtypevar () in
+    Fun (Typevar new_tv, (true, false), Ref (Typevar new_tv))
+  ;;
+
+  let ref_f = Variable (ref_t, "ref")
+
+  let deref_t =
+    let new_tv = newtypevar () in
+    Fun (Ref (Typevar new_tv), (false, false), Typevar new_tv)
```

Comment:
It is non-intuitive but reading from a reference should also be considered an effectful operation. Consider the case `let r = ref 0 in ((r := 1), !r)`: the result depends on the evaluation order because `r := 1` and `!r` are simultaneous "effects". If something depends on the global mutable state, you should think of it as effectful.

Reply:
Fixed. Would also like your feedback on `update_t` in `effast.ml`. If `(:=)` partially applied to a ref, its effect bits are (false, false), right? The effect is only present when the second argument (new value) is passed to the fn.

Reply:
Yes, the effects on this form of currying are correct (and quite subtle, well done!).

Reply:
(You mention this is fixed, but I still see the old definition in the PR. Could it be that you forgot to push an updated version?)